Add Delete blocks (#144)
axiomhq/docs: restapi/delete-blocks.mdx

---

> title: 'Delete blocks'
> description: 'This page explains how to delete blocks of data.'

Author: ```suggestion
title: 'Delete blocks of data'
description: 'This page explains how to delete blocks of data.'
sidebarTitle: 'Delete blocks'
```

---

> title: 'Delete blocks of data'

Reviewer: I would expect the details of this endpoint to be in the Dataset endpoints section. What is the reason this does not feature in Swagger definition?

Author: We agreed that this endpoint is not ready for GA.

description: 'This page explains how to delete blocks of data.'
sidebarTitle: 'Delete blocks'
---

> After data ingestion, Axiom stores events in blocks of up to 65,536 events. You cannot delete individual events after you send them to Axiom, but you can delete entire blocks of data. With the `/v2/datasets/_apl/delete` endpoint, you can specify an APL query and delete all blocks that contain matching events.

Reviewer: "to 65,536 events" Do we need to be this specific (expect impl could change), or would "tens of thousands of events" be ok? I actually think there could be scenarios where it’s less than that (based on time), but we should explore that offline.

Author: Agreed! Suggested a change below.

> After data ingestion, Axiom stores events in blocks of up to 65,536 events. You cannot delete individual events after you send them to Axiom, but you can delete entire blocks of data. With the `/v2/datasets/_apl/delete` endpoint, you can specify an APL query and delete all blocks that contain matching events.

Author: ```suggestion
After data ingestion, Axiom stores events in blocks of tens of thousands of events. You cannot delete individual events after you send them to Axiom, but you can delete entire blocks of data. With the `/v2/datasets/_apl/delete` endpoint, you can specify an APL query and delete all blocks that contain matching events.
```

---
title: 'Delete blocks of data'
description: 'This page explains how to delete blocks of data.'
sidebarTitle: 'Delete blocks'
---

> After data ingestion, Axiom stores events in blocks of tens of thousands of events. You cannot delete individual events after you send them to Axiom, but you can delete entire blocks of data. With the `/v2/datasets/_apl/delete` endpoint, you can specify an APL query and delete all blocks that contain matching events.

Reviewer 2: > You cannot delete individual events after you send them to Axiom

_right now_ you cannot, that will hopefully change in the future

<Warning>

> The `/v2/datasets/_apl/delete` endpoint deletes all blocks containing events that match your query. For example, even if a block only contains one event that matches your query, the endpoint removes all 65,536 events in the block. For this reason, using this endpoint may remove more data than your query specifies.

Reviewer 2: > all 65,536 events

there could be anywhere from 1 to a few million events in a block

> Deleting data using this endpoint is slow and expensive. Only use this endpoint when you have sent sensitive data to Axiom by mistake and you need to immediately delete it for regulatory reasons.

Reviewer 2: > for regulatory reasons

do we need to specify reason they have for deletion?

> Use the block deletion endpoint for deleting less than one million rows in one request. To delete more than one million rows, send a number of requests and delete a limited number of rows in each request. If you delete too many rows in one request, the request might fail unexpectedly.

Reviewer 2: > might fail unexpectedly.

If we're telling them about it, is it unexpected?

</Warning>

## Delete blocks

To delete all blocks containing events that match an APL query, send a `POST` request to `/v2/datasets/_apl/delete`.

To authenticate your request, use an API token with update permissions for datasets. For more information, see [API tokens](/reference/tokens).

In the request body, include a JSON object with the following fields:

- `apl`: An APL query that specifies the events to be deleted. You can only use the operators [`where`](/apl/tabular-operators/where-operator) and [`search`](/apl/tabular-operators/search-operator).
- `startTime`: The start of the time period over which you wantto run the query.
- `endTime`: The end of the time period over which you want to run the query.
- `commit`: A Boolean that specifies whether to actually perform the deletion. Set `commit` to `false` to preview the impact of the deletion request without actually deleting any events.

A valid request receives a 200 response with the following fields present in the response body:

- `dryRun`: Returns `true` if the value of `commit` specified in the request was `false`.
- `rowsMatched`: The number of matching events.

> - `rowsDeleted`: The number of deleted events. This is the number of all events in all the blocks that contain events matching your query. The number of deleted events is higher than the number of matching events.

Reviewer 2: ```suggestion
- `rowsDeleted`: The number of deleted events. This is the number of all events in all the blocks that contain events matching your query. The number of deleted events can be higher than the number of matching events.
```

- `firstMatchedEvent`: The timestamp of the first event matching the query.
- `lastMatchedEvent`: The timestamp of the last event matching the query.

When you make a successful request with `commit` set to `true`, blocks containing matching events are queued for deletion. The action cannot be reversed. A block queued for deletion is typically deleted within 5 minutes, but it may take longer if your organization makes a large number of deletion requests.

## Best practices

> - Start with a dry run. Set `commit` to `false` to safely review which events are to be deleted by your request.

Reviewer 2: ```suggestion
- Start with a dry run. Set `commit` to `false` to safely review the number of events to be deleted by your request.
```
